src/Futhark/Internalise/NormaliseSize.hs

```diff
@@ -0,0 +1,570 @@
+-- | This Normalisation module converts a well-typed, monomorphised,
+-- module-free Futhark program into an equivalent program without
+-- arbitrary size expression.
+-- This is done by replacing all complex size expresssions by an equivalent
+-- of the old style NamedSize e.g. size just described as a variable.
+-- This variables are then calculated in let-bindings or as a size parameter,
+-- depending on the context.
+module Futhark.Internalise.NormaliseSize (transformProg) where
+
+import Control.Monad.Identity
+import Control.Monad.RWS hiding (Sum)
+import Control.Monad.Reader
+import Control.Monad.State
+import Control.Monad.Writer hiding (Sum)
+import Data.Bifunctor
+import Data.Foldable
+import Data.List (nub)
+import Data.Map.Strict qualified as M
+import Data.Maybe
+import Data.Set qualified as S
+import Futhark.MonadFreshNames
+import Futhark.Util (mapAccumLM)
+import Futhark.Util.Pretty
+import Language.Futhark
+import Language.Futhark.Traversals
+import Language.Futhark.TypeChecker.Types
+
+-- | Box for Exp that allows better comparison than straight forward structural equality
+newtype ReplacedExp = ReplacedExp {unReplaced :: Exp}
+  deriving (Show)
+
+instance Pretty ReplacedExp where
+  pretty (ReplacedExp e) = pretty e
+
+instance Eq ReplacedExp where
+  ReplacedExp e1 == ReplacedExp e2
+    | Just es <- similarExps e1 e2 =
+        all (uncurry (==) . bimap ReplacedExp ReplacedExp) es
+  _ == _ = False
+
+instance Ord ReplacedExp where
```

Comment:
I am somewhat uncomfortable with this instance, since it's not obvious to me that it is actually correct (e.g. transitive), which may break at some point.  Could we replace the `M.Map ReplacedExp VName` values with `[(ReplacedExp,VName)]` and use `Eq`-based list operations instead?  That of course involves far more linear searches, but I suspect these mappings are usually small, so there might not be a performance loss.